File: server/content.py
import re
from bs4 import BeautifulSoup
import requests

from server.utils import get_clean_url


def get_final_url(raw_url):
    url = get_clean_url(raw_url)
    redirectResponse = requests.get(url, allow_redirects=True, timeout=10)
    return redirectResponse.url


def get_clean_tags(soup):
    forbidden_tags = [
        "script",
        "style",
        "noscript",
        "iframe",
        "footer",
        "nav",
        "form",
        "svg",
        "link",
        "img",
        "head",
    ]
    allowed_tags = ["h1", "h2", "h3", "h4", "p", "ul", "ol", "code"]

    # remove unnecessary tags recursive
    for tag in soup(forbidden_tags):
        tag.decompose()

    # remove all attibutes
    for tag in soup.find_all(True):
        tag.attrs = {}  # type: ignore

    # combine string array
    return "".join(str(el) for el in soup.find_all(allowed_tags))


def get_html(soup: BeautifulSoup):
    raw_content = get_clean_tags(soup)

    # remove unnecessary code
    content = re.sub(r"<([a-zA-Z0-9]+)>\s*</\1>", "", raw_content)
    content = content.replace("\n", "")
    content = re.sub(r"\s{2,}", " ", content)

    return content


def get_text(soup: BeautifulSoup):
    raw_content = get_clean_tags(soup)
    soup = BeautifulSoup(raw_content, "html.parser")

    # Nur den reinen Text holen
    clean_text = soup.get_text(separator="\n", strip=True)

    return clean_text

File: server/src/content.ts
import { parseHTML } from "linkedom";
import { getCleanUrl } from "./utils.js";

export const getFinalUrl = async (rawUrl: string): Promise<string> => {
  const cleanUrl = getCleanUrl(rawUrl);
  const res = await fetch(cleanUrl, {
    method: "GET",
    redirect: "follow",
  });
  return res.url || cleanUrl;
};

export const extractContent = async (url: string) => {
  const res = await fetch(url);
  const html = await res.text();

  const { document } = parseHTML(html);

  const charset =
    document.querySelector("meta[charset]")?.getAttribute("charset") || "utf-8";

  const root =
    document.querySelector("article")?.innerHTML ||
    document.querySelector("main")?.innerHTML ||
    document.body?.innerHTML ||
    "";

  const { document: contentDoc } = parseHTML(root);

  const KlassesAndIds = [
    "footer",
    "ads",
    "banner",
    "popup",
    "cookie",
    "newsletter",
    "breadcrumb",
    "navigation",
    "menu",
    "hidden",
  ];

  const forbidden = [
    "head",
    "script",
    "link",
    "style",
    "noscript",
    "iframe",
    "nav",
    "footer",
    "form",
    "input",
    "svg",
    "img",
    "video",
    "progress",
    ...KlassesAndIds.flatMap((item) => [
      `[class*="${item}"]`,
      `[id*="${item}"]`,
    ]),
  ];

  forbidden.forEach((tag) => {
    contentDoc.querySelectorAll(tag).forEach((el) => el.remove());
  });

  // Schlüsselwörter, die im Tag-Namen enthalten sein dürfen, um ihn zu entfernen
  const forbiddenTagKeywords = ["header", "footer", "sidebar", "widget"];

  // Entferne alle Tags, deren Name eines der Keywords enthält
  contentDoc.querySelectorAll("*").forEach((el) => {
    const tagName = el.tagName.toLowerCase();
    if (
      forbiddenTagKeywords.some((keyword) => tagName.includes(`${keyword}-`))
    ) {
      el.remove();
    }
  });

  // Alle Attribute entfernen
  contentDoc.querySelectorAll("*").forEach((el) => {
    [...el.attributes].forEach((attr) => el.removeAttribute(attr.name));
  });

  const result = contentDoc.toString();

  const plainText = result
    .replace(/<[^>]*>/g, "")
    .replace(/\s+/g, " ")
    .replace(/\n+/g, "\n")
    .trim();

  return {
    meta: { title: document.title, charset },
    content: {
      html: result,
      plain: plainText,
    },
  };
};
